src/sys/unix/tcp.rs

```diff
@@ -1,101 +1,30 @@
+use crate::sys::Socket;
+
 use std::io;
-use std::mem::{size_of, MaybeUninit};
 use std::net::{self, SocketAddr};
-use std::os::unix::io::{AsRawFd, FromRawFd};
-
-use crate::sys::unix::net::{new_ip_socket, socket_addr, to_socket_addr};
+use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd};
 
 pub fn connect(addr: SocketAddr) -> io::Result<net::TcpStream> {
-    new_ip_socket(addr, libc::SOCK_STREAM)
-        .and_then(|socket| {
-            let (raw_addr, raw_addr_length) = socket_addr(&addr);
-            syscall!(connect(socket, raw_addr, raw_addr_length))
-                .or_else(|err| match err {
-                    // Connect hasn't finished, but that is fine.
-                    ref err if err.raw_os_error() == Some(libc::EINPROGRESS) => Ok(0),
-                    err => Err(err),
-                })
-                .map(|_| socket)
-                .map_err(|err| {
-                    // Close the socket if we hit an error, ignoring the error
-                    // from closing since we can't pass back two errors.
-                    let _ = unsafe { libc::close(socket) };
-                    err
-                })
-        })
-        .map(|socket| unsafe { net::TcpStream::from_raw_fd(socket) })
+    let socket = Socket::from_addr(addr, libc::SOCK_STREAM)?;
+
+    // Set SO_REUSEADDR (mirrors what libstd does).
+    socket.set_reuse_address()?;
+
+    socket.connect(addr)?;
+    unsafe { Ok(net::TcpStream::from_raw_fd(socket.into_raw_fd())) }
 }
 
 pub fn bind(addr: SocketAddr) -> io::Result<net::TcpListener> {
-    new_ip_socket(addr, libc::SOCK_STREAM).and_then(|socket| {
-        // Set SO_REUSEADDR (mirrors what libstd does).
-        syscall!(setsockopt(
-            socket,
-            libc::SOL_SOCKET,
-            libc::SO_REUSEADDR,
-            &1 as *const libc::c_int as *const libc::c_void,
-            size_of::<libc::c_int>() as libc::socklen_t,
-        ))
-        .and_then(|_| {
-            let (raw_addr, raw_addr_length) = socket_addr(&addr);
-            syscall!(bind(socket, raw_addr, raw_addr_length))
-        })
-        .and_then(|_| syscall!(listen(socket, 1024)))
-        .map_err(|err| {
-            // Close the socket if we hit an error, ignoring the error
-            // from closing since we can't pass back two errors.
-            let _ = unsafe { libc::close(socket) };
-            err
-        })
-        .map(|_| unsafe { net::TcpListener::from_raw_fd(socket) })
-    })
+    let socket = Socket::from_addr(addr, libc::SOCK_STREAM)?;
+    socket.bind(addr)?;
+    socket.listen(1024)?;
+    unsafe { Ok(net::TcpListener::from_raw_fd(socket.into_raw_fd())) }
```

Comment:
This pattern is quite risky: if any method fails between creating the Socket and casting it into another type, an FD will leak.
A good alternative may be implementing Drop for Socket so the fd gets closed no matter what. That will also reduce some code duplication :)

Reply:
The `Socket` methods this dispatches to will close the FD on errors. Here is `bind2` for [example](https://github.com/kleimkuhler/mio/blob/0c22d3d57d16ebefa0e75606c42b3190d8e7a6a7/src/sys/unix/socket.rs#L156-L161). This currently doesn't happen for `Socket::listen` so that should be added!

If the `bind` or `listen` then fail, the FD would be closed and it wouldn't try to create a `TcpListener`.

Reply:
Well, as I said IMO it's too risky, because whatever new code that will use this interface in the future won't be aware of this implicit "critical section" in which FDs may leak.
So `Socket::from_addr` should either return a fully safe-to-use socket, or a clearly WIP-dont-drop-me-now socket.

Reply:
Oh okay yes I see what you mean. Yep that is good point. I mentioned in my comment below that there may be some further discussion if this is even a change Mio wants to consider, but I can add a `Drop` impl for `Socket` to reduce some code duplication and ensure a leak doesn't occur for methods where there isn't an explicit `libc::close`.